ee/agent/storage/bbolt/backup.go

```diff
@@ -0,0 +1,234 @@
+		)
+		return nil
+	case <-time.After(backupInitialDelay):
+		break
+	}
+
+	// Take periodic backups
+	ticker := time.NewTicker(backupInterval)
+	defer ticker.Stop()
+	for {
+		if err := d.backupDb(); err != nil {
+			d.slogger.Log(context.TODO(), slog.LevelWarn,
+				"could not perform periodic database backup",
+				"err", err,
+			)
+		}
+
+		select {
+		case <-ticker.C:
+			continue
+		case <-d.interrupt:
+			d.slogger.Log(context.TODO(), slog.LevelDebug,
+				"interrupt received, exiting execute loop",
+			)
+			return nil
+		}
+	}
+}
+
+func (d *databaseBackupSaver) Interrupt(_ error) {
+	// Only perform shutdown tasks on first call to interrupt -- no need to repeat on potential extra calls.
+	if d.interrupted {
+		return
+	}
+	d.interrupted = true
+
+	d.interrupt <- struct{}{}
+}
+
+func (d *databaseBackupSaver) backupDb() error {
+	// Take backup in temporary location
+	backupLocation := BackupLauncherDbLocation(d.knapsack.RootDirectory())
+	tempBackupLocation := fmt.Sprintf("%s.tmp", backupLocation)
+	defer func() {
+		// In case we errored out when taking the backup, clean up the temp file
+		_ = os.Remove(tempBackupLocation)
+	}()
+
+	if err := d.knapsack.BboltDB().View(func(tx *bbolt.Tx) error {
+		return tx.CopyFile(tempBackupLocation, 0600)
+	}); err != nil {
+		return fmt.Errorf("backing up database: %w", err)
+	}
+
+	// Confirm file exists and is nonempty
+	if exists, err := nonEmptyFileExists(tempBackupLocation); !exists {
+		return fmt.Errorf("backup succeeded, but nonempty file does not exist at %s", tempBackupLocation)
+	} else if err != nil {
+		return fmt.Errorf("backup succeeded, but error checking if file was created at %s: %w", tempBackupLocation, err)
+	}
+
+	// Perform rotation of older backups so we can move this backup to `backupLocation`
+	if err := d.rotate(); err != nil {
+		return fmt.Errorf("backup succeeded, but rotation did not: %w", err)
+	}
+
+	if err := os.Rename(tempBackupLocation, backupLocation); err != nil {
+		return fmt.Errorf("renaming temp backup %s to %s after rotation: %w", tempBackupLocation, backupLocation, err)
+	}
+
+	// Log success
+	d.slogger.Log(context.TODO(), slog.LevelDebug,
+		"took backup",
+		"backup_location", backupLocation,
+	)
+
+	return nil
+}
+
+func (d *databaseBackupSaver) rotate() error {
+	baseBackupPath := BackupLauncherDbLocation(d.knapsack.RootDirectory())
+
+	for i := numberOfOldBackupsToRetain; i > 0; i -= 1 {
+		currentBackupPath := fmt.Sprintf("%s.%d", baseBackupPath, i)
+
+		// This backup doesn't exist yet -- skip it
+		if _, err := os.Stat(currentBackupPath); err != nil && os.IsNotExist(err) {
+			continue
+		}
+
+		// If is the oldest backup, delete it so we can rotate a new one into its place
+		if i == numberOfOldBackupsToRetain {
+			if err := os.Remove(currentBackupPath); err != nil {
+				return fmt.Errorf("removing oldest backup %s during rotation: %w", currentBackupPath, err)
+			}
+			continue
+		}
```

Comment:
You can probably drop this logic, by removing the `i`th backup before the loop, and looping `i := numberOfOldBackupsToRetain - 1`.

Reply:
Updated! Hopefully it reads a little better now